### PR TITLE
feat: aleo-snap-adapter

### LINKED ISSUE #33: adapter

### PR DESCRIPTION
**Pull request by irubido**
Closes #33

### CODE REVIEW

File: packages/adapter/package.json [hunk 1]
```diff
@@ -16,5 +18,10 @@
     "@types/node": "18.15.11",
     "eslint": "^8.37.0",
     "typescript": "^5.0.3"
+  },
+  "dependencies": {
+    "@chainsafe/aleo-snap": "workspace:^",
+    "@chainsafe/aleo-snap-types": "workspace:^",
+    "@chainsafe/aleo-snap-wasm": "workspace:^"
```
Comment (BeroBurny) [inline packages/adapter/package.json:25]: Please remove `@chainsafe/aleo-snap-wasm` as `dependencies` there should not be any usages of wasm in adapter

File: packages/adapter/src/utils.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    return !!Object.values(await getWalletSnaps()).find(
+      (permission) =>
+        permission.id === snapOrigin &&
+        (!version || permission.version === version)
+    );
+  } catch (e) {
+    console.log("Failed to obtain installed snaps", e);
+    return false;
+  }
+}
```
Comment (BeroBurny) [inline packages/adapter/src/utils.ts:40]: please use `web3_clientVersion` to figure out if is flask

File: packages/types/package.json [hunk 3]
```diff
@@ -0,0 +1,17 @@
+{
+  "name": "@chainsafe/aleo-snap-types",
```
Comment (BeroBurny) [inline packages/types/package.json:2]: Seems this package is better to be called shared, so can have code for the adapter and snap

**Review by BeroBurny — CHANGES_REQUESTED** [review]

File: .yarn/releases/yarn-3.6.0.cjs
Comment (BeroBurny) [inline .yarn/releases/yarn-3.6.0.cjs:1]: remove this file

File: .yarnrc.yml [hunk 4]
```diff
@@ -15,3 +15,5 @@ plugins:
     spec: "@yarnpkg/plugin-typescript"
   - path: .yarn/plugins/@yarnpkg/plugin-engines.cjs
     spec: "https://raw.githubusercontent.com/devoto13/yarn-plugin-engines/main/bundles/%40yarnpkg/plugin-engines.js"
+
+yarnPath: .yarn/releases/yarn-3.6.0.cjs
```
Comment (BeroBurny) [inline .yarnrc.yml:19]: remove this

File: packages/adapter/package.json [hunk 5]
```diff
@@ -21,7 +21,6 @@
   },
   "dependencies": {
     "@chainsafe/aleo-snap": "workspace:^",
```
Comment (BeroBurny) [inline packages/adapter/package.json:23]: there is no point using aleo-snap in adapter

File: packages/adapter/src/utils.ts [hunk 6]
```diff
@@ -38,3 +21,28 @@ export async function isSnapInstalled(
...
+    throw new Error("Metamask is not installed");
+  }
+  return await Promise.race([
+    isMetaMaskFlask(),
+    new Promise<boolean>(() =>
+      setTimeout(() => {
+        return false;
+      }, 1000)
+    ),
+  ]);
```
Comment (BeroBurny) [inline packages/adapter/src/utils.ts:47]: can you add a comment on why we need a race condition here?

File: packages/snap/package.json [hunk 7]
```diff
@@ -58,8 +58,8 @@
     "typescript": "^5.0.3"
   },
   "dependencies": {
-    "@chainsafe/aleo-snap-types": "workspace:^",
+    "@chainsafe/aleo-snap-shared": "workspace:^",
     "@chainsafe/aleo-snap-wasm": "workspace:^",
-    "@metamask/snaps-ui": "^0.32.2"
+    "@metamask/snaps-ui": "^0.3types2.2"
```
Comment (BeroBurny) [inline packages/snap/package.json:63]: this version seems to have a typo, dint find this one on [npm](https://www.npmjs.com/package/@metamask/snaps-ui?activeTab=versions)

File: packages/snap/test/integration/general.test.ts [hunk 8]
```diff
@@ -2,8 +2,8 @@ import { expect, use } from "chai";
 import sinonChai from "sinon-chai";
 import chaiAsPromised from "chai-as-promised";
 import { Dappeteer, DappeteerPage, initSnapEnv } from "@chainsafe/dappeteer";
-import { Methods } from "../../src";
 import { buildSnap } from "./utils";
+import { Methods } from "../../../types";
```
Comment (BeroBurny) [inline packages/snap/test/integration/general.test.ts]: please use package import, do not leave the workspace package to access some code on another package

File: packages/types/index.d.ts [hunk 9]
```diff
@@ -1,5 +1,7 @@
 import { Signature } from "@chainsafe/aleo-snap-wasm";
 
+export { Signature } from "@chainsafe/aleo-snap-wasm";
```
Comment (BeroBurny) [inline packages/types/index.d.ts]: is this duplicated?

**Review by BeroBurny — CHANGES_REQUESTED** [review]

File: packages/shared/index.ts [hunk 10]
```diff
@@ -0,0 +1,78 @@
+import { Signature } from "@chainsafe/aleo-snap-wasm";
+
+export { Signature } from "@chainsafe/aleo-snap-wasm";
```
Comment (BeroBurny) [inline packages/shared/index.ts:3]: seems this import is duplicated

File: packages/shared/package.json [hunk 11]
```diff
@@ -0,0 +1,23 @@
+{
+  "name": "@chainsafe/aleo-snap-shared",
+  "version": "0.0.1",
+  "types": "./index.ts",
+  "license": "(Apache-2.0 AND MIT)",
```
Comment (BeroBurny) [inline packages/shared/package.json:5]: there is missing `main` field

File: packages/snap/src/index.ts [hunk 12]
```diff
@@ -8,16 +8,16 @@ import { sign } from "./rpc/sign";
...
-export enum Methods {
+let wasm: InitOutput;
+
+enum Methods {
   GetAccount = "aleo_getAccount",
   GetViewKey = "aleo_getViewKey",
   Decrypt = "aleo_decrypt",
   Sign = "aleo_sign",
   Verify = "aleo_verify",
 }
```
Comment (BeroBurny) [inline packages/snap/src/index.ts:19]: Use shared package for enum

**Review by BeroBurny — CHANGES_REQUESTED** [review]

File: packages/shared/index.ts (hunk 10 above)
Comment (irubido) [reply packages/shared/index.ts:3]: exporting Signature in shared, so that example doesn't have to have aleo-snap-wasm dependency

**Pull request merged**